src/datepicker/utils/date-helpers.js

```diff
@@ -15,6 +25,19 @@ class DateHelpers<T> {
   constructor(adapter: DateIOAdapter<T>) {
     this.adapter = adapter;
   }
+  getAdapterWithNewLocale: mixed => DateIOAdapter<T> = locale => {
```

Comment:
This was the biggest blocker in this PR.  The problem is that the `date-io` instance that we're passing in as a prop accepts a `locale` when its created, but there are helpers that are reliant on passing locale as an argument. 

This means that we either need a way to change the locale for the instance, or we need to recreate the instance with a new locale.  As a short term solution, I settled on recreating the instance.  Doing that requires the `adapterOptionMap` above, which makes it possible to us to get the options passed into the `Utils` class when it was instantiated, so we can create it over again with a new locale. 

The thing is, I really don't like this solution, so I [submitted a PR](https://github.com/dmtrKovalenko/date-io/issues/347) to see if they make some changes on their end to make this process a little bit easier to manage.  I just didn't want to have to stop making progress until that problem is solved.

Reply:
That makes a ton of sense.  What I'll do is push forward on making the `Datepicker` changes(which is probably going to take me a while) and if I can get a solution from `date-io` in the meantime, I'll apply that, otherwise, I'll add the deprecation warnings.